crates/storage/provider/src/transaction.rs

```diff
@@ -913,73 +958,39 @@ where
-                                        .map(|(key, val)| (U256::from_be_bytes(key.0), val))
-                                        .collect(),
-                                    wipe_storage: false, /* it is always false as all storage
-                                                          * changesets for selfdestruct are
-                                                          * already accounted. */
-                                },
-                            )
-                        })
-                        .collect();
-                    next_changeset = changeset_iter.next().unwrap_or_default();
-                    changeset
-                } else {
-                    BTreeMap::new()
-                };
-
+                if let Some(changes) = all_changesets.remove(&next_transition_id) {
+                    for mut change in changes.into_iter() {
+                        change
+                            .set_transition_id(block_exec_res.transitions_count() as TransitionId);
+                        block_exec_res.add_and_apply(change);
+                    }
+                }
+                block_exec_res.finish_transition();
                 next_transition_id += 1;
-                block_exec_res.tx_changesets.push(TransactionChangeSet {
-                    receipt: Receipt::default(), /* TODO(receipt) when they are saved, load them
-                                                  * from db */
-                    changeset,
-                    new_bytecodes: Default::default(), /* TODO(bytecode), bytecode is not cleared
-                                                        * so it is same sa previous. */
-                });
             }
 
             let Some((_,block_transition)) = block_transition_iter.next() else { break};
             // if block transition points to 1+next transition id it means that there is block
             // changeset.
             if block_transition == next_transition_id + 1 {
-                // assert last_transition_id == block_transition
-                if next_transition_id == next_changeset.0 {
-                    // take block changeset
-                    block_exec_res.block_changesets = next_changeset
-                        .1
-                        .into_iter()
-                        .map(|(address, (account, _))| (address, account))
-                        .collect();
-                    next_changeset = changeset_iter.next().unwrap_or_default();
+                if let Some(changes) = all_changesets.remove(&next_transition_id) {
+                    for mut change in changes.into_iter() {
+                        change
+                            .set_transition_id(block_exec_res.transitions_count() as TransitionId);
+                        block_exec_res.add_and_apply(change);
+                    }
+                    block_exec_res.finish_transition();
+                    next_transition_id += 1;
                 }
```

Comment:
Should we log an `error!` if this was None?

Reply:
Don't think it is even possible that it is none as all transactions will always result in at least one change, but I am unsure. I just kept it in this structure to make it work for now, I will refactor it later (it was way too much to mentally keep track of in its current state)